src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,715 @@
+
+            public abstract List<TSource> ToList();
+
+            public abstract int GetCount(bool onlyIfCheap);
+        }
+
+        private sealed class Append1Iterator<TSource> : AppendPrependIterator<TSource>
+        {
+            private readonly TSource _element;
+
+            public Append1Iterator(IEnumerable<TSource> source, TSource element)
+                : base(source)
+            {
+                _element = element;
+            }
+
+            public override Iterator<TSource> Clone() => new Append1Iterator<TSource>(_source, _element);
+
+            public override bool MoveNext()
+            {
+                switch (_state)
+                {
+                    case 1:
+                        GetSourceEnumerator();
+                        _state = 2;
+                        goto case 2;
+                    case 2:
+                        if (!LoadFromEnumerator())
+                        {
+                            _current = _element;
+                            _state = -1;
+                        }
+                        return true;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override AppendPrependIterator<TSource> Append(TSource element) => new AppendNIterator<TSource>(_source, _element, element);
+
+            public override AppendPrependIterator<TSource> Prepend(TSource element) => new AppendPrependNIterator<TSource>(_source, new SharedCollection<TSource>(element), 1, new SharedCollection<TSource>(_element), 1);
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + 1;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + 1 : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                int index = 0;
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - 1);
+                array[index] = _element;
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(4) : new List<TSource>(count);
```

Comment:
Why are you explicitly specifying 4?  Let's just let List use its default heuristics, which will start with an empty array and grow to 4 the first time something is added.